File: software/raspberry_pi/src/navegation/navegation_4.0/sensors.py
from pymavlink import mavutil
from time import time
import math as m
from AUVError import CollisionDetected, PixhawkHighTempError, RaspberryHighTempError

# conferir as conversões
CONV_TO_MS = 9.80665 / 1000  # convert mG to m/s²
CONV_TO_RAD = 1 / 1000   # convert mrad/s to rad/s
CONV_TO_T = 1 / 10000000 # convert mgauss to Tesla

# Acceleration limit to consider a collision
ACC_LIMIT = 15

# Critical temperatures
MAX_TEMP_PIXHAWK = 50
MAX_TEMP_RASPBERRY = 85
MAX_TEMP_JETSON = 60

class Sensors:
    def __init__(self):
        """
        Initializes the Sensors class, establishing a connection to Pixhawk and setting initial values 
        for acceleration, gyro, magnetometer, velocity, and temperature sensors.
        """

        # serial connetion ('/dev/serial0', baud = 57600)
        # intranet connetion ('127.0.0.1:12550)
        self.connection = mavutil.mavlink_connection('/dev/serial0', baud = 57600)
        self.acc = [0, 0, 0] # current acceleration [x, y, z]
        self.gyro = [0, 0, 0] # [x, y, z]
        self.mag = [0, 0, 0] # [x, y, z]
        self.vel = [0, 0, 0] # [x, y, z]
        self.temp_pixhawk = None
        self.temp_raspberry = None

        self.current_time = time()
        self.old_time = self.current_time

        self.__send_heartbeat()

    def update_data(self):
        """
        Continuously updates sensor data, fetching new information from Pixhawk and temperature sensors
        """

        while True:
            self.old_time = self.current_time

            self.__update_data_imu()
            self.__update_data_temp()

            self.current_time = time()

    def __send_heartbeat(self):
        """
        Sends the heartbeat to pixhawk for establish connection
        """
        self.connection.mav.heartbeat_send(mavutil.mavlink.MAV_TYPE_GCS, mavutil.mavlink.MAV_AUTOPILOT_INVALID, 0, 0, 0)
    
    def __update_data_imu(self):
        """
        updates acceleration, gyroscope, magnetometer and speed data
        """

        is_scaled_imu2_valid = 0
        is_global_position_int_valid = 0

        while not is_global_position_int_valid or not is_scaled_imu2_valid:
                msg = self.connection.recv_match(type=['GLOBAL_POSITION_INT', 'SCALED_IMU2'])

                if msg is None:
                    continue

                if msg.get_type() == 'SCALED_IMU2':
                    self.acc = [msg.xacc * CONV_TO_MS, msg.yacc * CONV_TO_MS, msg.zacc * CONV_TO_MS]

                    self.gyro = [msg.xgyro * CONV_TO_RAD, msg.ygyro * CONV_TO_RAD, msg.zgyro * CONV_TO_RAD]
                    # Dados de mag não funcionam
                    self.mag = [msg.xmag, msg.ymag, msg.zmag]
                    
                    is_scaled_imu2_valid = 1
                if msg.get_type() == 'GLOBAL_POSITION_INT':
                    self.vel = [msg.vx / 100, msg.vy / 100, msg.vz / 100]

                    is_global_position_int_valid = 1

    # IMPLEMENTAR
    def __update_data_temp(self):
        """
        Updates system temperature data
        """

        pass

    def collision_detect(self):
        """
        Detects whether the AUV has crashed based on acceleration data from Pixhawk
        """

        acceleration = self.get_acc()

        if any(m.fabs(acc) > ACC_LIMIT for acc in acceleration):
            raise CollisionDetected(acceleration, self.current_time)
    
    # TERMINAR
    def detect_overheat(self):
        """
        checks whether the systems temperature is safe
        """

        if self.temp_pixhawk > MAX_TEMP_PIXHAWK:
            raise PixhawkHighTempError(self.temp_pixhawk, self.current_time)
        if self.temp_raspberry > MAX_TEMP_RASPBERRY:
            raise RaspberryHighTempError(self.temp_raspberry, self.current_time)
    
    def get_acc(self):
        return self.acc
    
    def get_gyro(self):
        return self.gyro
    
    def get_mag(self):
        return self.mag
    
    def get_vel(self):
        return self.vel

File: software/raspberry_pi/src/navegation/navegation_5.0/sensors.cpp
#include <array>
#include <chrono>
#include <thread>
#include <iostream>
#include <mavsdk/mavsdk.h>
#include <mavsdk/plugins/telemetry/telemetry.h>
#include <mavsdk/plugins/mavlink_passthrough/mavlink_passthrough.h>
#include "auv_error.cpp"

using namespace std;
using namespace chrono;
using namespace mavsdk;
using namespace this_thread;

// Convertions
double CONV_TO_MS = .01;       // convert cm/s to m/s 
double CONV_TO_MS2 = 9806.65;  // convert mG to m/s²
double CONV_TO_RAD = .001;     // convert mrad/s to rad/s
double CONV_TO_µT = .1;        // convert mgauss to µT

//Acceleration limit consider a collision
double ACC_LIMIT = 15;

// Critical temperatures - !!!Determinar quais são as temperaturas maximas!!!
double MAX_TEMP_PIXHAWK = 60;
double MAX_TEMP_RASPBERRY = 70;
double MAX_TEMP_JETSON = 60;

class Sensors{
    private:
        array<double, 3> acc = {0, 0, 0};       // {x, y, z}
        array<double, 3> gyro = {0, 0, 0};      // {x, y, z}
        array<double, 3> mag = {0, 0, 0};       // {x, y, z}
        array<double, 3> vel = {0, 0, 0};       // {x, y, z}
        double tempPixhawk;                     // ºC
        double tempRaspberry;                   // ºC
        steady_clock::time_point currentTime;
        steady_clock::time_point oldTime;
        Telemetry *telemetry;
        MavlinkPassthrough *mavlink_passthrough;

    public:
        /**
         * @brief Initializes the Sensors class, establishing a connection to Pixhawk and setting initial values for acceleration, gyro, magnetometer,
         *  velocity, and temperature sensors.
         */
        Sensors(){
            Mavsdk mavsdk{Mavsdk::Configuration{Mavsdk::ComponentType::GroundStation}};

            ConnectionResult connection_result = mavsdk.add_any_connection("serial:///dev/ttyACM0:115200");

            if (connection_result != ConnectionResult::Success){ //Colocar isso em um código de erro
                cout << "Failed to connect: " << connection_result << endl;
            } else cout << "Connected to Pixhawk" << endl;

            sleep_for(seconds(2));

            auto systems = mavsdk.systems();

            if(!systems.empty()) {
                auto system = mavsdk.systems().at(0);
                
                this->telemetry = new Telemetry{system};
                this->mavlink_passthrough = new MavlinkPassthrough{system};

                while (!this->telemetry->health().is_accelerometer_calibration_ok ||
                    !this->telemetry->health().is_gyrometer_calibration_ok || 
                    !this->telemetry->health().is_magnetometer_calibration_ok) {
                    std::cout << "Waiting for the system to be ready..." << std::endl;
                    sleep_for(seconds(1));
                }
    
                std::cout << "System is ready!" << std::endl;
            }else {//Colocar isso em código de erro tbm
                std::cerr << "Failed to detect system." << std::endl;
            }
        }

        /**
         * @brief Continuously updates sensor data, fetching new information from Pixhawk and ****temperature sensors (fazer essa parte)****
         */
        void updateData(){
            cout << "Listening messages..." << endl;

            this->mavlink_passthrough->subscribe_message(MAVLINK_MSG_ID_GLOBAL_POSITION_INT, [this](const mavlink_message_t& message) {
                mavlink_global_position_int_t imu_data;
                mavlink_msg_global_position_int_decode(&message, &imu_data);

                this->vel[0] = imu_data.vx * CONV_TO_MS;
                this->vel[1] = imu_data.vy * CONV_TO_MS;
                this->vel[2] = imu_data.vz * CONV_TO_MS;
            });

            this->mavlink_passthrough->subscribe_message(MAVLINK_MSG_ID_RAW_IMU, [this](const mavlink_message_t& message) {
                mavlink_raw_imu_t imu_data;
                mavlink_msg_raw_imu_decode(&message, &imu_data);

                this->acc[0] = imu_data.xacc * CONV_TO_MS2;
                this->acc[1] = imu_data.yacc * CONV_TO_MS2;
                this->acc[2] = imu_data.zacc * CONV_TO_MS2;

                this->gyro[0] = imu_data.xgyro * CONV_TO_RAD;
                this->gyro[1] = imu_data.ygyro * CONV_TO_RAD;
                this->gyro[2] = imu_data.zgyro * CONV_TO_RAD;

                this->mag[0] = imu_data.xmag * CONV_TO_µT;
                this->mag[1] = imu_data.ymag * CONV_TO_µT;
                this->mag[2] = imu_data.zmag * CONV_TO_µT;

                this->tempPixhawk = imu_data.temperature / 100.0;
            });

            while (1) sleep_for(milliseconds(100));
        }

        /**
         * @brief Detects whether the AUV has crashed based on acceleration data from Pixhawk
         */
        void collisionDetect(){
            if(fabs(this->acc[0]) > ACC_LIMIT || fabs(this->acc[1]) > ACC_LIMIT || fabs(this->acc[2]) > ACC_LIMIT){
                throw CollisionDetected(acc);
            }
        }

        /**
         * @brief Checks whether the systems temperature is safe **** Terminar!! ****
         */
        void detectOverheat(){
            if(this->tempPixhawk > MAX_TEMP_PIXHAWK) throw PixhawkHighTemperature(tempPixhawk);
            if(this->tempRaspberry > MAX_TEMP_RASPBERRY) throw RaspberryHighTemperature(tempRaspberry);
        }

        void finish(){
            delete this->telemetry;
            delete this->mavlink_passthrough;
        }

        array<double, 3> getAcc(){
            return this->acc;
        }

        array<double, 3> getGyro(){
            return this->gyro;
        }

        array<double, 3> getMag(){
            return this->mag;
        }

        array<double, 3> getVel(){
            return this->vel;
        }
};